src/wasm-builder.h

```diff
@@ -1328,8 +1328,8 @@ class Builder {
     block->name = name;
     if (append) {
       block->list.push_back(append);
-      block->finalize(type);
     }
+    block->finalize(type);
```

Comment:
Drive-by fix. We should finalize even if `append` is nullptr

Reply:
Is that true? If we aren't appending anything, then should the block already exist and be finalized or be brand new and finalized by `makeBlock`?

Reply:
I think the old logic here makes sense in that it assumes that if there is a previous block then that block is valid (i.e. finalized), but the comment for the function was not that clear (probably my fault).

@aheejin is this change needed for the new pass for some reason? If so perhaps we can add a new function or rethink this one.

Reply:
I think you're right, but things are a little tricky.

So first off, I think we should've passed `name` to `makeBlock` in the first place, because `makeBlock` uses `BranchSeeker` when finalizing only when the block has a name, and we need that `BranchSeeker` to figure out the inner `try_table (catch_ref $label)` is targeting `$label` and its return type is `exnref`.
https://github.com/WebAssembly/binaryen/blob/185019919b1fa2efe8d02056827a31f4f50dd01e/src/wasm/wasm.cpp#L197-L209

But there's another problem. If the passed expression is already a block, `blockifyWithName` reuses it, in which case the passed `type` parameter (in case it is not null) doesn't get used. This doesn't mean the existing block is invalid. For example, in `processDelegateTarget`, the input to `blockfiyWithName` is
```wast
(block
  (try_table (catch_all_ref $label)
    ...
  )
)
```
The `block` here is the body of the outer `try` (which is targeted by the inner `delegate`, which has been converted to `try_table`). The naive output we would want is this:
```wast
(block $label (result exnref)
  (block
    (try_table (catch_all_ref $label)
      ...
    )
  )
)
```
`(block $label)`'s result type is `exnref` because it is targeted by `catch_all_ref`. But with `blockifyWithName`, we can reuse the input block if it is already a block. So we can do
```wast
(block $label (result exnref)
  (try_table (catch_all_ref $label)
    ...
  )
)
```
But in this case we need a refinalization of the `block` with the given type `type`, even if the given block is not invalid, because we are reusing it.

I updated `blockifyWithName` so that it satisfies this, PTAL. So now we pass `name` to `makeBlock`, and also do finalization when `type` is explicitly given.

```diff
--- a/src/wasm-builder.h
+++ b/src/wasm-builder.h
@@ -1323,11 +1323,14 @@ public:
       block = any->dynCast<Block>();
     }
     if (!block || block->name.is()) {
-      block = makeBlock(any);
+      block = makeBlock(name, any);
+    } else {
+      block->name = name;
     }
-    block->name = name;
     if (append) {
       block->list.push_back(append);
+    }
+    if (append || type) {
       block->finalize(type);
     }
     return block;
```

Reply:
Done in #6223, even though I'm not sure how to add a standalone test there... (Will rebase this PR onto that after that lands, because I'm not using any fancy tools supporting stacked PRs)